### PR TITLE
fix consensus algorithm

### LINKED ISSUE #642: Consensus bug: invalid chain reorganization
### _Type_:
Issue

### _What happened? / What do you need?_:
It happens quite a few times during e2e. And it can cause the reorg chain error:
`ERROR[07-12|09:24:11.003] Impossible reorg, please file an issue oldnum=26 oldhash=e2441a…15148c newnum=26 newhash=e2441a…15148c`

### _If it's a request, what do you need it for?_:
To have a valid consensus.

### _Priority_:
High

### PR DESCRIPTION
**Pull request by JekaMas**
Mandatory issues are:
 * we always do 'reorg' blockchain operation for every received block
 * we choose as a new blockchain any received valid chain
 * we don't check does a new block is already in the blockchain

Closes #642

### CODE REVIEW

**Comment by yourheropaul** [comment]
I think let's maybe get a couple more reviews before merging.

**Comment by JekaMas** [comment]
@yourheropaul agree! The more review is done, the better.

**Comment by JekaMas** [comment]
@yourheropaul How do you think, can we merge this PR?

File: client/core/blockchain.go [hunk 1]
```diff
@@ -902,21 +909,39 @@ func (bc *BlockChain) WriteBlockWithState(block *types.Block, receipts []*types.
 
 	rawdb.WriteReceipts(batch, block.Hash(), block.NumberU64(), receipts)
 
-	// Reorganise the chain if the parent is not the head block
-	if block.ParentHash() != currentBlock.Hash() {
-		if err := bc.reorg(currentBlock, block); err != nil {
-			return NonStatTy, err
+	//todo: @jeka try to use cumulative votes amount
```
Comment (jmatosp) [inline client/core/blockchain.go]: hard to tell what this is doing, can you use functions that describe these operations

ex: 
`isSameBlockHigh`
`isBlockHigher`
`needsReorg`
`isHeadBlock`
...

File: client/core/headerchain.go [hunk 2]
```diff
@@ -124,36 +124,45 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
 
 	rawdb.WriteHeader(hc.chainDb, header)
 
-	// Delete any canonical number assignments above the new head
-	batch := hc.chainDb.NewBatch()
-	for i := number + 1; ; i++ {
-		hash := rawdb.ReadCanonicalHash(hc.chainDb, i)
-		if hash == (common.Hash{}) {
-			break
+	//todo: @jeka try to use cumulative votes amount
```
Comment (jmatosp) [inline client/core/headerchain.go]: no todo's, add a issue instead

**Comment by JekaMas** [comment]
It's a bit scarry, but allthe tests pass each time. :D

**Comment by JekaMas** [comment]
@acroca After last dev merge tests have started to be unstable. Can it be 'mint' changes?

**Comment by acroca** [comment]
@JekaMas I'm not sure, so far I've seen only the one I've reported in issue #650
If you find more issues let me know and I'll try to find out.

**Comment by JekaMas** [comment]
@rgeraldes Could you also review the issue?

**Pull request merged**